riscv-sbi.adoc

```diff
@@ -0,0 +1,174 @@
+(C) 2019 Palmer Dabbelt <palmer@sifive.com>
+(C) 2019 Atish Parta <atish.parta@wdc.com>
+
+It is licensed under the Creative Commons Attribution 4.0 International
+License (CC-BY 4.0).  The full license text is available at
+https://creativecommons.org/licenses/by/4.0/.
+
+## Introduction
+
+This specification describes the RISC-V Supervisor Binary Interface, known from
+here on as SBI.  This interface allows supervisor-mode software to be written
+that is portable to all RISC-V implementations.  The design of the SBI follows
+the general RISC-V philosophy of having a small core along with a set of
+optional modular extensions.
+
+### Versioning and History
+
+This document describes a draft of version 0.2 of the RISC-V SBI specification.
+
+#### Changes Since Version 0.1
+
+* The entire v0.1 SBI has been moved to the legacy extension, which is now an
+  optional extension.  This is technically a backwards-incompatible change, 
+
+## Binary Encoding
+
+All SBI functions share a single binary encoding, which facilitates the mixing
+of SBI extensions.  This binary encoding matches the standard RISC-V UNIX
+syscall ABI, which itself is based on the calling convention defined in the
+RISC-V ELF psABI.  In other words, SBI calls are exactly the same as standard
+RISC-V function calls except that:
+
+* An `ecall` is used as the control transfer instruction instead of a `call`
+  instruction.
+* `a7` (or `t0` on RV32E-based systems) encodes the SBI extension ID, which
+  matches how the system call ID is encoded in the standard UNIX system call
+  ABI.
+
+Many SBI extensions also chose to encode an additional function ID in `a6`,
+a scheme similar to the `ioctl()` system call on many UNIX operating systems.
+This allows SBI extensions to encode multiple functions within the space of a
+single extension.
+
+In the name of compatibility, SBI extension IDs and SBI function IDs are
+encoded as signed 32-bit integers.  When passed in registers these follow the
+standard RISC-V calling convention rules.
+
+SBI functions generally return a pair of values in `a0` and `a1`, with `a1`
+returning an error code.  This is analogous to returning the C structure
+
+    struct sbiret {
+        long value;
+        long error;
+    };
+
+Standard SBI error codes are listed below
+
+| Error Type               | Value  |
+| -------------------------|:------:|
+|  SBI_SUCCESS             |  0     |
+|  SBI_ERR_FAILURE         | -1     |
+|  SBI_ERR_NOT_SUPPORTED   | -2     |
+|  SBI_ERR_INVALID_PARAM   | -3     |
+|  SBI_ERR_DENIED          | -4     |
+|  SBI_ERR_INVALID_ADDRESS | -5     |
+
+
+## SBI Base Functionality, Extension ID 0x10
+
+The base of the supervisor binary interface is designed to be as small as
+possible.  As such, it only contains functionality for probing which SBI
+extensions are available and for querying the version of the SBI.  All
+functions in the base must be supported by all SBI implementations, so there
+are no error returns defined.
```

Comment:
Is a1 always set to SBI_SUCCESS by these functions, or do they leave a1 in some other/unspecified state?